Source/Dafny/Dafny.atg

```diff
@@ -867,131 +867,162 @@ DeclModifier<ref DeclModifierData dmod>
           .)
     )
+    .
 
-  | (.
-      IToken exportId;
-      List<ExportSignature> exports = new List<ExportSignature>();;
-      List<IToken> extends = new List<IToken>();
-      bool provideAll = false;
-      bool revealAll = false;
-      bool isDefault = false;
-      bool isRefining = false;
-      ExportSignature exsig;
-    .)
-    "export" (. exportId = t; .)
-    [ ExportIdent<out exportId> ]
-    [ "..."  (. isRefining = true; .) ]
-    {
-      "provides"
-      (
-        ( ModuleExportSignature<true, out exsig>        (. exports.Add(exsig); .)
-          {"," ModuleExportSignature<true, out exsig>   (. exports.Add(exsig); .) }
-        )
-      | "*" (. provideAll = true; .)
+/*------------------------------------------------------------------------*/
+ExportId<out IToken id> = NoUSIdentOrDigits<out id> .
+
+ModuleExport<ModuleDefinition parent, out ModuleDecl submodule>
+= "export"
+  (.
+    IToken exportId = t;
+    IToken/*!*/ id;
+    List<ExportSignature> exports = new List<ExportSignature>();;
+    List<IToken> extends = new List<IToken>();
+    bool provideAll = false;
+    bool revealAll = false;
+    bool isDefault = false;
+    bool isRefining = false;
+    ExportSignature exsig;
+  .)
+  [ ExportId<out exportId> ]
+  [ "..."  (. isRefining = true; .) ]
+  {
+    "provides"
+    (
+      ( ExportSignature<true, out exsig>        (. exports.Add(exsig); .)
+        {"," ExportSignature<true, out exsig>   (. exports.Add(exsig); .) }
       )
-    | "reveals"
-      (
-        ( ModuleExportSignature<false, out exsig>       (. exports.Add(exsig); .)
-          {"," ModuleExportSignature<false, out exsig>  (. exports.Add(exsig); .) }
-        )
-      | "*" (. revealAll = true; .)
+    | "*" (. provideAll = true; .)
+    )
+  | "reveals"
+    (
+      ( ExportSignature<false, out exsig>       (. exports.Add(exsig); .)
+        {"," ExportSignature<false, out exsig>  (. exports.Add(exsig); .) }
       )
-    | "extends"
-      ExportIdent<out id>       (. extends.Add(id); .)
-      {"," ExportIdent<out id>  (. extends.Add(id); .) }
-    }
-    (. if (exportId.val == "export" || exportId.val == parent.Name) {
-         isDefault = true;
-       }
-       submodule = new ModuleExportDecl(exportId, parent, exports, extends, provideAll, revealAll, isDefault, isRefining);
-    .)
-  )
-.
+    | "*" (. revealAll = true; .)
+    )
+  | "extends"
+    ExportId<out id>       (. extends.Add(id); .)
+    {"," ExportId<out id>  (. extends.Add(id); .) }
+  }
+  (. if (exportId.val == "export" || exportId.val == parent.Name) {
+       isDefault = true;
+     }
+     submodule = new ModuleExportDecl(exportId, parent, exports, extends, provideAll, revealAll, isDefault, isRefining);
+  .)
+  .
+
+/*------------------------------------------------------------------------*/
+// TODO - this is somewhat a hack because it relies on TypeNameOrCtorSuffix
```

Comment:
Is there a way to track this (maybe even a GitHub issue)? Want to reduce the chance of this being incorrect in the future.

Reply:
I'll attempt a fix when I address types in the next round of edits